lib/matplotlib/colorbar.py

```diff
@@ -1035,14 +1035,11 @@ def remove(self):
         except AttributeError:
             return
         try:
-            gs = ax.get_subplotspec().get_gridspec()
-            subplotspec = gs.get_topmost_subplotspec()
-        except AttributeError:
-            # use_gridspec was False
+            subplotspec = self.ax.get_subplotspec().get_gridspec()._subplot_spec
```

Comment:
Can you walk through how this works?  Where did `_subplot_spec` come from and why is it defined on the parent gridspec?

Reply:
The classic (tighr_layout, pre-CL) way of positioning colorbars (assuming that the parent axes is positioned with a subplotspec) is to create a subgridspec on the parent subplotspec.  For example say we had initially
```
A B

C D
```
and we choose to add a colorbar to axes D (which was previously positioned at `gs = GridSpec(2, 2)[1, 1])`.  We create a subgridspec `sgs = GridSpec(2, 2)[1, 1].subgridspec(3, 2)` so that the block where we previously had D is replace by (zoomed in)
```
D .
D c
D .
```
where D is now at `sgs[:, 1]` and c (the colorbar) at `sgs[1, 1]` (and likewise for other colorbar locations).
Now if we remove the colorbar, we want to move back D to `gs`, which is exactly equal to `sgs._subplot_spec` (\*); `sgs` itself is equal to `c.get_subplotspec().get_gridspec()`.

The code prior to this PR did not go up one nesting layer, but rather climbed all of the nesting layers indiscriminately (get_topmost_subplotspec), which would be problematic if the whole thing way already layered into other nested subgridspecs.

(\*) as noted in the message, maybe a public API for _subplot_spec would be nice, but that's another issue.

Reply:
I see - `_subplot_spec` is the subplot spec of the (Sub)GridSpec.  Makes sense.